src/Compilers/CSharp/Portable/Lowering/ExtensionMethodReferenceRewriter.cs

```diff
@@ -149,6 +146,61 @@ static RefKind argumentRefKindFromReceiverRefKind(RefKind receiverRefKind)
             }
         }
 
-        // PROTOTYPE: public override BoundNode? VisitDelegateCreationExpression(BoundDelegateCreationExpression node)
+        [return: NotNullIfNotNull(nameof(method))]
+        private static MethodSymbol? VisitMethodSymbolWithExtensionRewrite(BoundTreeRewriter rewriter, MethodSymbol? method)
+        {
+            if (method is { OriginalDefinition.ContainingSymbol: NamedTypeSymbol { IsExtension: true } declaringTypeDefinition } &&
+                method.OriginalDefinition.TryGetCorrespondingExtensionImplementationMethod() is MethodSymbol implementationMethod)
+            {
+                method = implementationMethod.AsMember(method.ContainingSymbol.ContainingType).
+                    ConstructIfGeneric(method.ContainingType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.Concat(method.TypeArgumentsWithAnnotations));
+            }
+
+            return rewriter.VisitMethodSymbol(method);
+        }
+
+        [return: NotNullIfNotNull(nameof(method))]
+        public override MethodSymbol? VisitMethodSymbol(MethodSymbol? method)
+        {
+            Debug.Assert(method is not { OriginalDefinition.ContainingSymbol: NamedTypeSymbol { IsExtension: true } } ||
+                         method.OriginalDefinition.TryGetCorrespondingExtensionImplementationMethod() is null);
+            return base.VisitMethodSymbol(method);
+        }
+
+        public override BoundNode? VisitDelegateCreationExpression(BoundDelegateCreationExpression node)
+        {
+            return VisitDelegateCreationExpression(this, node);
+        }
+
+        public static BoundNode VisitDelegateCreationExpression(BoundTreeRewriter rewriter, BoundDelegateCreationExpression node)
+        {
+            return updateDelegateCreation(node, VisitMethodSymbolWithExtensionRewrite(rewriter, node.MethodOpt), (BoundExpression)rewriter.Visit(node.Argument), rewriter.VisitType(node.Type));
+
+            static BoundNode updateDelegateCreation(BoundDelegateCreationExpression node, MethodSymbol? methodOpt, BoundExpression argument, TypeSymbol type)
```

Comment:
nit: consider inlining so that the `Update` call is at the top-level or change the local function returns `isExtensionMethod` value. I'm happy to do it for you in a follow-up PR if it's any inconvenience and you're okay with it